pytorch_lightning/trainer/trainer.py

```diff
@@ -1431,58 +1437,100 @@ def _call_teardown_hook(self) -> None:
-                # todo: move this data parallel logic into the data parallel plugin
-                output = accelerator_output if output is None else output
-
-            # call the ttp hook
-            if hook_name not in ("setup", "teardown", "on_train_start") and hasattr(
-                self.training_type_plugin, hook_name
-            ):
-                ttp_hook = getattr(self.training_type_plugin, hook_name)
-                ttp_output = ttp_hook(*args, **kwargs)
-                output = ttp_output if output is None else output
+            fn = getattr(pl_module, hook_name)
+            if callable(fn):
+                # TODO: when profiling separate hook name by LM hooks/Callback hooks
+                with self.profiler.profile(hook_name):
+                    output = fn(*args, **kwargs)
+
+            # restore current_fx when nested context
+            pl_module._current_fx_name = prev_fx_name
+
+        return output
+
+    def _call_callback_hooks(
+        self,
+        hook_name: str,
+        *args: Any,
+        **kwargs: Any,
+    ) -> Optional[Any]:
+        pl_module = self.lightning_module
+        output = None
+
+        if pl_module:
+            prev_fx_name = pl_module._current_fx_name
+            pl_module._current_fx_name = hook_name
+
+        # TODO: remove code in if statement in v1.7
+        if hook_name in ("on_train_batch_start", "on_train_batch_end"):
+            fn = getattr(self, hook_name)
+            if callable(fn):
+                with self.profiler.profile(hook_name):
+                    output = fn(*args, **kwargs)
+        else:
+            for callback in self.callbacks:
+                if hook_name in ("on_init_start", "on_init_end"):
+                    # these `Callback` hooks are the only ones that do not take a lightning module.
+                    # we also don't profile bc profiler hasn't been set yet
+                    fn = getattr(callback, hook_name)
+                    if callable(fn):
+                        output = fn(self, *args, **kwargs)
+                else:
+                    fn = getattr(callback, hook_name)
+                    if callable(fn):
+                        with self.profiler.profile(hook_name):
+                            output = fn(self, self.lightning_module, *args, **kwargs)
 
         if pl_module:
             # restore current_fx when nested context
             pl_module._current_fx_name = prev_fx_name
 
         return output
 
+    # TODO: eventually no longer need this
+    def _call_ttp_hook(
+        self,
+        hook_name: str,
+        *args: Any,
+        **kwargs: Any,
+    ) -> Optional[Any]:
+        output = None
+        fn = getattr(self.training_type_plugin, hook_name)
+        if callable(fn):
+            with self.profiler.profile(hook_name):
+                output = fn(*args, **kwargs)
+
+        return output
+
+    # TODO: eventually no longer need this
+    def _call_accelerator_hook(
```

Comment:
Do you see a need for call connectors hook ?

Reply:
Hmm I think it's better to leave the "call hook" functions to user defined hooks? The "hooks" in the connectors are not user defined right?

Reply:
Yes